File: turbo/apps/desktop/src/window-policy.ts
type WindowOpenDecision =
  | { readonly action: "allow-in-app" }
  | { readonly action: "open-external"; readonly url: string }
  | { readonly action: "deny" };

const EXTERNAL_PROTOCOLS = new Set(["http:", "https:", "mailto:"]);

function parseUrl(rawUrl: string): URL | null {
  try {
    return new URL(rawUrl);
  } catch {
    return null;
  }
}

export function isAllowedAppNavigation(
  rawUrl: string,
  allowedAppOrigins: ReadonlySet<string>,
): boolean {
  const url = parseUrl(rawUrl);
  if (!url) {
    return false;
  }
  return allowedAppOrigins.has(url.origin);
}

export function decideWindowOpen(
  rawUrl: string,
  allowedAppOrigins: ReadonlySet<string>,
): WindowOpenDecision {
  const url = parseUrl(rawUrl);
  if (!url) {
    return { action: "deny" };
  }

  if (allowedAppOrigins.has(url.origin)) {
    return { action: "allow-in-app" };
  }

  if (EXTERNAL_PROTOCOLS.has(url.protocol)) {
    return { action: "open-external", url: url.toString() };
  }

  return { action: "deny" };
}

File: turbo/apps/desktop/native/OkouDesktop/Sources/OkouDesktopKit/Core/WindowPolicy.swift
import Foundation

public enum WindowOpenDecision: Equatable, Sendable {
    case allowInApp
    case openExternal(url: String)
    case deny
}

/// Navigation policy shared by every desktop window. Port of `window-policy.ts`.
public enum DesktopWindowPolicy {
    static let externalSchemes: Set<String> = ["http", "https", "mailto"]

    public static func isAllowedAppNavigation(_ rawUrl: String, allowedAppOrigins: Set<String>) -> Bool {
        guard let origin = DesktopURL.origin(rawUrl) else { return false }
        return allowedAppOrigins.contains(origin)
    }

    public static func decideWindowOpen(_ rawUrl: String, allowedAppOrigins: Set<String>) -> WindowOpenDecision {
        guard let url = DesktopURL.parse(rawUrl) else {
            return .deny
        }
        if let origin = DesktopURL.origin(url), allowedAppOrigins.contains(origin) {
            return .allowInApp
        }
        if externalSchemes.contains(DesktopURL.scheme(url)) {
            return .openExternal(url: url.string ?? rawUrl)
        }
        return .deny
    }
}
